core/diagnostics.rs

```diff
@@ -0,0 +1,871 @@
+  pub fn from_compiler_json_value(v: &serde_json::Value) -> Option<Self> {
+    if !v.is_object() {
+      return None;
+    }
+    let obj = v.as_object().unwrap();
+
+    // required properties
+    let message_v = &obj["message"];
+    if !message_v.is_string() {
+      return None;
+    }
+    let message = String::from(message_v.as_str().unwrap());
+    let category = DenoDiagnosticCategory::from_i64(
+      obj.get("category").and_then(Value::as_i64).unwrap(),
+    );
+    let source = DenoDiagnosticSources::from_i64(
+      obj.get("source").and_then(Value::as_i64).unwrap(),
+    );
+
+    let source_line = obj
+      .get("sourceLine")
+      .and_then(|v| v.as_str().map(String::from));
+    let script_resource_name = obj
+      .get("scriptResourceName")
+      .and_then(|v| v.as_str().map(String::from));
+
+    let code = obj.get("code").and_then(Value::as_i64);
+    let line_number = obj.get("lineNumber").and_then(Value::as_i64);
+    let start_position = obj.get("startPosition").and_then(Value::as_i64);
+    let end_position = obj.get("endPosition").and_then(Value::as_i64);
+    let start_column = obj.get("startColumn").and_then(Value::as_i64);
+    let end_column = obj.get("endColumn").and_then(Value::as_i64);
+
+    let diagnostics_v = obj.get("diagnostics");
+    let diagnostics = match diagnostics_v {
+      Some(d) => {
+        let mut diagnostics = Vec::<DenoDiagnostic>::new();
+        let diagnostic_values = d.as_array().unwrap();
+
+        for diagnostic_v in diagnostic_values {
+          match DenoDiagnostic::from_compiler_json_value(diagnostic_v) {
+            Some(diagnostic) => diagnostics.push(diagnostic),
+            None => return None,
+          }
+        }
+
+        Some(diagnostics)
+      }
+      _ => None,
+    };
+
+    let related_information_v = obj.get("relatedInformation");
+    let related_information = match related_information_v {
+      Some(r) => {
+        let mut related_information = Vec::<DenoDiagnostic>::new();
+        let related_info_values = r.as_array().unwrap();
+
+        for related_info_v in related_info_values {
+          match DenoDiagnostic::from_compiler_json_value(related_info_v) {
+            Some(related_info) => related_information.push(related_info),
+            _ => return None,
+          }
+        }
+
+        Some(related_information)
+      }
+      _ => None,
+    };
+
+    let next_v = obj.get("next");
+    let next = match next_v {
+      Some(n) => Some(Box::new(
+        DenoDiagnostic::from_compiler_json_value(n).unwrap(),
+      )),
+      _ => None,
+    };
+
+    Some(DenoDiagnostic {
+      message,
+      diagnostics,
+      related_information,
+      code,
+      source_line,
+      script_resource_name,
+      line_number,
+      start_position,
+      end_position,
+      category,
+      source,
+      start_column,
+      end_column,
+      frames: None,
+      next,
+    })
+  }
+
+  pub fn from_json_value(v: serde_json::Value) -> Option<Self> {
```

Comment:
This also should have a unit test.

Reply:
Oooops, I forgot to add this one, it isn't my code, it was from `core/js_error.rs` originally and doesn't have a unit test there, as it is from the V8 error, but I will add one.